### ISSUE TITLE
Kafka Pub/Sub documentation should note consumer group behaviour when subscribing to multiple topics

### ISSUE THREAD

**Issue by KrylixZA** [dapr/docs#3707]
**Describe the issue**
When using Dapr to subscribe to multiple topics in Kafka, it should be noted on the documentation that the consumer group which is created will be distributed across all the topics. 

Say, for instance, you are subscribing to two topics: one with 10 partitions, and another with 30. You may presume, in order to scale to your absolute maximum for performance, you would want to run 40 pods; with 10 pods being assigned to the first topic, 30 to the other. However, this is not always the case (but could be). There is no guarantee about how consumers are balanced a cross the topic.

The consumer group will do its best to balance consumers, but it can't know for sure what the balance should be. This becomes particularly challenging if you have fairly peaky volumes traversing through your topics, and you use some kind of autoscaler to attempt to match the load, you might end up with pods not consuming and consumer lag growing.

The solution to my problem has been to configure multiple Pub/Sub components, with different names, different consumer groups, all pointing to the same Kafka broker. Ultimately it comes down to spinning up separate consumer groups. Using the default consumer group of `sarama` or even setting the consumer group name to `{appID}` (and possibly `{namespace}` if you only have one namespace) will replicate this problem. For example, my two Pub/Sub components are configured as follows:

```yaml
- name: consumerGroup
  value: "{appID}-topic-one"
- name: clientID 
  value: "{appID}"
```

```yaml
- name: consumerGroup
  value: "{appID}-topic-two"
- name: clientID 
  value: "{appID}"
```

**URL of the docs**
https://docs.dapr.io/reference/components-reference/supported-pubsub/setup-apache-kafka/

**Expected content**
The docs should mention this behaviour and inform users of ways to get around this particular problem. The simplest being multiple Pub/Sub components. This effectively creates two connections to the broker, under separate consumer groups, distributing pods to consumers exactly 1:1.

**Screenshots**
N/A

**Additional context**
N/A

**Comment by msfussell** [comment]
@KrylixZA - Would you be open to submitting a suggested PR for this change? Thanks

**Comment by iliaspoli** [comment]
we have ran upon this issue as well, is this a kafka specific behavior or Dapr is doing it?
we would expect messages to be distributed per topic in parallel not one by one across all topics...

**Comment by KrylixZA** [comment]
@IliasP91, it's technically not Dapr. The behaviour is driven by limitations within the [IBM/sarama](https://github.com/IBM/sarama) library. You can follow this issue for the specific behaviour: https://github.com/IBM/sarama/issues/2621

To work around this, for every service that consumes from more than one topic in Kafka, we've introduced an individual Dapr Pub/Sub component for each topic and used that while subscribing. We then have a specific consumer group name defined within that Dapr component YAML to make it unique. This forces all pods/containers running within a single consumer group to be fully allocated to that topic. It's not ideal, but it works. 

Just remember to make sure that your maximum scaling properties are set against the topic with the highest partition count and if you are using something like KEDA to scale against the topic with the highest throughput.

**Comment by iliaspoli** [comment]
> @IliasP91, it's technically not Dapr. The behaviour is driven by limitations within the [IBM/sarama](https://github.com/IBM/sarama) library. You can follow this issue for the specific behaviour: [IBM/sarama#2621](https://github.com/IBM/sarama/issues/2621)
> 
> To work around this, for every service that consumes from more than one topic in Kafka, we've introduced an individual Dapr Pub/Sub component for each topic and used that while subscribing. We then have a specific consumer group name defined within that Dapr component YAML to make it unique. This forces all pods/containers running within a single consumer group to be fully allocated to that topic. It's not ideal, but it works.
> 
> Just remember to make sure that your maximum scaling properties are set against the topic with the highest partition count and if you are using something like KEDA to scale against the topic with the highest throughput.

I've seen the workaround and it would solve it but its not a good solution by far, our service in this case has more than 10 topic subscriptions

looking further into this does the following not suggest it has been fixed on dapr v1.11.3?
we do use 1.11.4 however and we are experiencing this problem

**Comment by KrylixZA** [comment]
Unfortunately it turned out that the issue was not completely fixed with Dapr 1.11.3. Our initial tests looked good but didn't quite materialise.

Agreed, the workaround is suboptimal at best. But it does work. We also have many streams but have had no choice but to implement it as such.

**Comment by iliaspoli** [comment]
> Unfortunately it turned out that the issue was not completely fixed with Dapr 1.11.3. Our initial tests looked good but didn't quite materialise.
> 
> Agreed, the workaround is suboptimal at best. But it does work. We also have many streams but have had no choice but to implement it as such.

Oh I see, that is very unfortunate, do you know if it is still broken on v1.12.x?

**Comment by KrylixZA** [comment]
Yeah same issue on 1.12. We have upgraded and unfortunately recently had to push a release to add more PubSub components to solve this very problem.
